### PR TITLE
Version 3.2

### PR DESCRIPTION
**Pull request by OpenSourceSimon**
The RedditVideoMakerBot has been updated to version 3.2, bringing exciting new features and bug fixes. Introducing **ElevenLabs TTS** for high-quality audio. Enjoy **background audio** and **random voices per comment**. Run the bot with one click using the **bat file**. Use the **zoom setting** for bigger text. Bug fixes include video downloader improvements, random story mode fix, updated dependencies, code optimizations, text size adjustments, enhanced Reddit credentials validation, and translator fixes. Enjoy the new features and **thanks to all contributors**!

### CODE REVIEW

File: main.py [hunk 1]
```diff
@@ -53,7 +54,11 @@ def main(POST_ID=None) -> None:
     bg_config = get_background_config()
     download_background(bg_config)
     chop_background_video(bg_config, length, reddit_object)
-    make_final_video(number_of_comments, length, reddit_object, bg_config)
+    try:
+        make_final_video(number_of_comments, length, reddit_object, bg_config)
+    except ffmpeg.Error as e:
+        print(e.stderr.decode("utf8"))
+        exit(1)
```
Comment (electro199) [inline main.py:61]: please keep this function clean

File: utils/.config.template.toml [hunk 2]
```diff
@@ -33,20 +33,20 @@ resolution_w = { optional = false, default = 1080, example = 1440, explantation
...
+voice_choice = { optional = false, default = "streamlabspolly", options = ["streamlabspolly", "tiktok", "googletranslate", "awspolly", "pyttsx", ], example = "tiktok", explanation = "The voice platform used for TTS generation. This can be left blank and you will be prompted to choose at runtime." }
 aws_polly_voice = { optional = false, default = "Matthew", example = "Matthew", explanation = "The voice used for AWS Polly" }
 streamlabs_polly_voice = { optional = false, default = "Matthew", example = "Matthew", explanation = "The voice used for Streamlabs Polly" }
 tiktok_voice = { optional = true, default = "en_us_001", example = "en_us_006", explanation = "The voice used for TikTok TTS" }
-tiktok_sessionid = { optional = true, example = "c76bcc3a7625abcc27b508c7db457ff1", explanation = "TikTok sessionid needed for the TTS API request. Check documentation if you don't know how to obtain it." }
+tiktok_sessionid = { optional = true, example = "c76bcc3a7625abcc27b508c7db457ff1", explanation = "TikTok sessionid needed if you're using the TikTok TTS. Check documentation if you don't know how to obtain it." }
```
Comment (electro199) [inline utils/.config.template.toml:49]: whose session-id is that

File: video_creation/background.py [hunk 3]
```diff
@@ -54,7 +53,7 @@ def get_background_config():
 
     # Handle default / not supported background using default option.
     # Default : pick random from supported background.
-    if not choice or choice not in background_options:
+    if not choice or choice not in background_options or choice == "random":
```
Comment (electro199) [inline video_creation/background.py:56]: why `list()`

File: main.py (hunk 1 above)
Comment (OpenSourceSimon) [reply main.py:61]: I agree with you, but this is the easiest way

File: video_creation/final_video.py [hunk 4]
```diff
@@ -177,6 +177,35 @@ def make_final_video(
...
+                    ffmpeg.input("assets/backgrounds/background.mp3")
+                    .filter(
+                        "volume",
+                        settings.config["settings"]["background"]["background_audio_volume"],
+                    )
+                )
+                # merges audio and bg_audio
+                merged_audio = ffmpeg.filter([audio, bg_audio], "amix", duration="longest")
+                # sets final audio to merged audio
+                audio = merged_audio
```
Comment (electro199) [inline video_creation/final_video.py:208]: make this in separate func

**Review by electro199 — CHANGES_REQUESTED** [review]

File: main.py (hunk 1 above)
Comment (electro199) [reply main.py:61]: i think by just watching where this exception can be raised and putting this try-except would work

File: utils/.config.template.toml (hunk 2 above)
Comment (OpenSourceSimon) [reply utils/.config.template.toml:49]: One that expired

File: video_creation/final_video.py (hunk 4 above)
Comment (OpenSourceSimon) [reply video_creation/final_video.py:208]: Doesn't matter

Comment (electro199) [reply video_creation/final_video.py:208]: this function is getting too big

File: TTS/engine_wrapper.py [hunk 5]
```diff
@@ -55,9 +55,20 @@ def add_periods(
...
+            comment["comment_body"] = re.sub(r'\bAI\b', 'A.I', comment["comment_body"])
+            comment["comment_body"] = re.sub(r'\bAGI\b', 'A.G.I', comment["comment_body"])
             if comment["comment_body"][-1] != ".":
                 comment["comment_body"] += "."
+            comment["comment_body"] = comment["comment_body"].replace(". . .", ".")
+            comment["comment_body"] = comment["comment_body"].replace(".. . ", ".")
+            comment["comment_body"] = comment["comment_body"].replace(". . ", ".")
+            comment["comment_body"] = re.sub(r'\."\.', '".', comment["comment_body"])
+            print(comment["comment_body"])
+
```
Comment (electro199) [inline TTS/engine_wrapper.py:71]: Why added processing and `print()` here remove print and move preprocessing code to block to `process_text`

File: TTS/engine_wrapper.py [hunk 6]
```diff
@@ -55,9 +55,20 @@ def add_periods(
         self,
     ):  # adds periods to the end of paragraphs (where people often forget to put them) so tts doesn't blend sentences
         for comment in self.reddit_object["comments"]:
+            # remove links
+            regex_urls = r"((http|https)\:\/\/)?[a-zA-Z0-9\.\/\?\:@\-_=#]+\.([a-zA-Z]){2,6}([a-zA-Z0-9\.\&\/\?\:@\-_=#])*"
+            comment["comment_body"] = re.sub(regex_urls, " ", comment["comment_body"])
```
Comment (electro199) [inline TTS/engine_wrapper.py:60]: urls are already cleaned

File: video_creation/background.py [hunk 7]
```diff
@@ -3,29 +3,38 @@
...
+        pos = background_options["video"][name][3]
 
-# Add position lambda function
-# (https://zulko.github.io/moviepy/ref/VideoClip/VideoClip.html#moviepy.video.VideoClip.VideoClip.set_position)
-for name in list(background_options.keys()):
-    pos = background_options[name][3]
+        if pos != "center":
+            background_options["video"][name][3] = lambda t: ("center", pos + t)
+
+    return background_options
```
Comment (electro199) [inline video_creation/background.py:36]: should we remove this @JasonLovesDoggo  it was used by moviepy now it is not needed

**Comment by OpenSourceSimon** [comment]
Progress on the update: it'll be 3.2 with a lot more features and bug fixes!

File: TTS/engine_wrapper.py (hunk 6 above)
Comment (OpenSourceSimon) [reply TTS/engine_wrapper.py:57]: I know, it's just to make sure

**Pull request merged**